cql3/selection/selection.cc

```diff
@@ -386,17 +388,61 @@ std::unique_ptr<result_set> result_set_builder::build() {
     return std::move(_result_set);
 }
 
-result_set_builder::restrictions_filter::restrictions_filter(::shared_ptr<restrictions::statement_restrictions> restrictions,
+namespace {
+
+using namespace expr;
+
+/// A visitor that copies an expression, but replaces token operator and unselected-column operators with the
+/// TRUE literal.
+struct drop_token_and_unselected_columns {
```

Comment:
How feasible is it to combine the drop with the extraction of the primary addressing path (primary key invtervals)?

These two parts have to be kept in sync, otherwise we can have something dropped that isn't taken care of by the other side (missing data) or not drop something from both paths (efficiently loss). IOW we must keep it either-or.

Something like

struct query_plan {
    expression primary_key_lookup;
    expression filter;
};

query_plan split_expression_to_primary_ley_lookup_and_filter(expression);

Reply:
Interesting question but irrelevant to do_filter().  This code only deals with post-fetch filtering, so the primary addressing path is irrelevant.  Its only aim is to ensure correct filtering of rows already fetched.

Reply:
But how does it know which parts of the expression are safe to drop? I think it only knows it by relying on knowledge of how the primary addressing path works.

Reply:
Correct. Perhaps it would be more logical to do this split upstream and only pass the filter here. Let me see how hard that is to code...

Reply:
I'm fine with not doing everything at once. What I want is
 - understand the path to enlightenment
 - add comments wherever we're doing something strange/weird/suboptimal, even if we're just keeping the status quo of the OO-style code